### PR TITLE
958 Unrecognized Argument on negative number queries

### LINKED ISSUE #958: %sql produces an error with a leading space and `-`
### What happens?

When a space precedes `-`, as in a numerical operation like `something * -1` , jupysql incorrectly parses the `-` to indicate a switch operator.

I presume this is the same root issue as #918, but this particular edge case was not addressed in #938

image

### To Reproduce

Run:
```
%sql select 5 * -1;
```

### OS:

Any

### JupySQL Version:

0.10.4

### Full Name:

Maciej Bukczynski

### Affiliation:

Darkhorse Analytics

### PR DESCRIPTION
**Pull request by bryannho**
## Describe your changes
- In `parse.py`, changed `split_args_and_sql()` to run anytime a SQL command is found, rather than only when JSON arrow operators are used
- This avoids corner cases where `-` is used to indicate a negative number in `SELECT` queries
- Added some spaces to `magic_args` to improve readability
- Added tests to verify correctness
- Added no-changelog label since this is a corner case of the other PR #918 

## Issue number

Closes #958 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
📚 Documentation preview 📚: https://jupysql--964.org.readthedocs.build/en/964/

### CODE REVIEW

**Comment by bryannho** [comment]
@neelasha23 @edublancas ready for review, confused why check for broken links is failing because the link seems to work for me. all other tests check out.

File: src/sql/parse.py [hunk 1]
```diff
@@ -259,9 +259,12 @@ def split_args_and_sql(line):
     """
     arg_line, sql_line = line, ""
 
-    # Only separate when json arrow operators used
-    # Otherwise, behavior is unchanged
-    if not any(op in line for op in ["->", "-->"]):
+    # Filenames may include sql keywords, so we omit them
+    check = re.sub(r"('.*')", "", line)
+    check = re.sub(r'(".*")', "", check)
```
Comment (edublancas) [inline src/sql/parse.py:264]: can you explain the logic of these two lines as a comment? It's unclear what this is doing

**Review by edublancas — CHANGES_REQUESTED** [review]

**Comment by edublancas** [comment]
@bryannho don't worry about the failing CI job, we can ignore it

File: src/tests/test_parse.py [hunk 2]
```diff
@@ -844,9 +844,9 @@ def test_escape_string_slicing_notation(query, expected_escaped, expected_found)
             "",
         ),
         (
-            "select * from authors",
             "select * from authors",
             "",
+            "select * from authors",
```
Comment (neelasha23) [inline src/tests/test_parse.py:849]: what exactly changed here?

File: src/tests/test_magic.py [hunk 3]
```diff
@@ -2597,3 +2597,19 @@ def test_table_does_not_exist_with_snippet_error(
...
     assert error_type == excinfo.value.error_type
     assert error_message in str(excinfo.value)
+
+
+@pytest.mark.parametrize(
+    "query",
+    [
+        "%sql select 5 * -2",
+        "%sql select 5 * - 2",
+        "%sql select 5 * -2;",
```
Comment (neelasha23) [inline src/tests/test_magic.py:2607]: you can also add some test case like numeric column of a test table * (some negative constant).  You can use the test `number_table`

File: src/sql/parse.py (hunk 1 above)
Comment (neelasha23) [reply src/sql/parse.py:264]: Please add the specific use cases also for which this logic was added. (json operators, spacing before negative sign, etc)

Comment (bryannho) [reply src/sql/parse.py]: [Added explanation in comments here.](https://github.com/ploomber/jupysql/pull/964/files#diff-22cfc6ec3e879941494b36e0dffaf24c100ab50173c413d821ebab94d0627ddfR262-R271)

The background behind this is in order to fix the original bug, I changed the logic for `split_args_and_sql` to split the line into an `args_line` and `sql_line` ANY time the line includes a SQL keyword. Previously it would only split the line if it included json operators `->` or `-->`, which left too many corner cases open. 

While testing this change, I noticed that queries like `%sqlplot boxplot --table 'file with spaces.csv'` were being split unintentionally, resulting in errors. This happened because 'file with spaces.csv' contains the SQL keyword `with`. Thus, I had to implement logic that would ignore any filenames when deciding if the line needed to be split, so that filenames could include words like `with`, `select`, or `begin` without being mistaken as actual SQL.

File: src/tests/test_parse.py (hunk 2 above)
Comment (bryannho) [reply src/tests/test_parse.py:849]: I changed the logic for `split_args_and_sql` to split any line that includes a SQL keyword. Previously it would only split the line if it included json operators `->` or `-->`. For this test case, it was previously NOT split. But with the changes I made, it now should be split.

File: src/tests/test_magic.py (hunk 3 above)
Comment (bryannho) [reply src/tests/test_magic.py]: [Added these test cases](https://github.com/ploomber/jupysql/pull/964/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2611-R2638)

File: src/sql/parse.py [hunk 4]
```diff
@@ -259,9 +259,22 @@ def split_args_and_sql(line):
     """
     arg_line, sql_line = line, ""
 
-    # Only separate when json arrow operators used
-    # Otherwise, behavior is unchanged
-    if not any(op in line for op in ["->", "-->"]):
+    # When queries include filenames, they may include SQL keywords
+    #   ex. 'penguins_selected'.csv contains "select"
+    # In these cases, splitting the query leads to parsing errors.
+    # So we ignore any filenames by removing text between double quotes ""
```
Comment (edublancas) [inline src/sql/parse.py:265]: I understand this now but I'm confused about why this changed was made. the issue reported from the user is about negative numbers, how is that connected to filenames?

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/sql/parse.py (hunk 4 above)
Comment (bryannho) [reply src/sql/parse.py:265]: The user reported that when negative numbers were included in the query, they triggered the same errors that `->` or `-->` did. This is because the parser thinks the `-` character is indicating an argument.

Previously, `split_args_and_sql` would conduct a split only when **both** of these conditions were met:
1. either `->` or `-->` was found in the query
2. any SQL command like `select` was found in the query

This was a bit too specific because it ignored corner cases like negative numbers. I decided to remove the first condition from `split_args_and_sql` so that it would conduct a split _every time_ a SQL command was found in the query, without checking for the arrow operators. So any `%sql` query containing the word `select`, `with`, etc. would now be split.

This change introduced a small bug, where any query containing a word like `select` or `with` in a **filename** would also be split. e.g. `%sqlplot boxplot --table 'file with spaces.csv'`.
This caused errors, because these words weren't acting as SQL commands/operators in this case. That's the reason I included [these lines](https://github.com/ploomber/jupysql/pull/964/files#diff-22cfc6ec3e879941494b36e0dffaf24c100ab50173c413d821ebab94d0627ddfR270-R271). While they aren't directly related to fixing the bug, they catch a specific case where the overall solution is a bit too liberal with its approach.

tldr; to solve this bug and possible future bugs, i made my solution more broad. this happened to be a bit _too_ broad, so I had to include a catch for one specific case (filenames).

I can see why you might feel like this is too confusing or specific. So if you would prefer, I can change it to be more similar to the original solution:
1. either `->`, `-->`, **OR a negative number** was found in the query
2. any SQL command like `select` was found in the query

I just thought that my new solution would be more generalizable in the case that there are even more corner cases that include `-` that we haven't stumbled upon yet. Let me know what you think! @edublancas

Comment (edublancas) [reply src/sql/parse.py:265]: thanks a lot for the detail explanation. I think the solution makes sense!

**Review by edublancas — APPROVED** [review]
